### DISCUSSION TITLE
Column does not exist on alias

### DISCUSSION THREAD

**Discussion by JasperVandenberghen** [mikro-orm/mikro-orm#4757] [Q&A] [answered]
**Describe the bug**

Each time I set an alias it will not recognize the column as existing. I have tried several ways of setting an alias but every time it says the same.

```
if (query.location && query.radius) {
      const [latitude, longitude] = query.location
      const srid = 4326
      const qb = this.activityRepository
        .qb('a')
        .select(`*`)
        .leftJoinAndSelect('organization', 'o')
        .leftJoin('o.users', 'ou')
        .leftJoin('ou.user', 'ouu')

      void qb
        .addSelect([
          `ST_Distance(ST_SetSRID(a.address_location, ${srid})::geography, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), ${srid})::geography)`,
        ])
        .as('distance')

      void qb.andWhere(
        `ST_DWithin(ST_SetSRID(a.address_location, ${srid})::geography, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), ${srid})::geography, ${
          query.radius * 1000
        }) = TRUE`
      )

      void qb.orderBy({ distance: QueryOrder.DESC })

      const count = await qb.getCount()

      void qb.limit(query.limit)
      void qb.offset(query.offset)

      const activities = await qb
      await this.em.populate(activities, ['categories', 'tags'])

      return [activities, count]
    }
```

**Stack trace**
```
error: order by "distance" desc limit 20 - column "distance" does not exist
```

**To Reproduce**
1. add a select with an alias
2. order by this alias

**Expected behavior**

It should recognize the alias column and order by it.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.12.1 |
| typescript | 4.7.4 |
| mikro-orm | 5.3.0 |

**Comment by B4nan** [comment]
The `qb.as()` method is for conversion of a QB instance to knex subquery instance, it does not do what you think - you don't alias the just selected column, but the whole QB (and you ignore the result of that method). You need to include the `as` part in your select:

```ts
qb.addSelect(`ST_Distance(ST_SetSRID(a.address_location, ${srid})::geography, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), ${srid})::geography) as distance`)
```

In v6 using `raw` or `sql` helper will be required (see #4197). Something like this should work in there:

```ts
qb.addSelect(sql.ref(`ST_Distance(ST_SetSRID(a.address_location, ${srid})::geography, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), ${srid})::geography)`).as('distance'))
```

**Reply by B4nan** [reply]
Btw you can also do this with a custom mapped type: https://mikro-orm.io/docs/custom-types#advanced-example---pointtype-and-wkt

**Reply by JasperVandenberghen** [reply]
I should have mentioned this but before I used the .as( ) function as a last resort I already had this:
```qb.addSelect(`ST_Distance(ST_SetSRID(a.address_location, ${srid})::geography, ST_SetSRID(ST_MakePoint(${longitude}, ${latitude}), ${srid})::geography) as distance`)``` which resulted in the same error.

**Reply by B4nan** [reply]
Well, it definitely didn't produce the same query - with that approach, you actually do alias the selected expression.

Also if you want to map that `distance` value to your entity, you will need a virtual property (`persist: false`) with that name defined on the entity, but that's another story.

**Reply by JasperVandenberghen** [reply]
Yes, I realize that now, although the correct alias throws the same error, but I don't see why.

I don't need to map the distance to my entity. It's just a temporary column used to order the results by distance.

**Reply by B4nan** [reply]
From what I just tried, you actually do need the virtual property, otherwise the `qb.orderBy` will throw as you try to order by an unknown property - it might be as well important for the QB to understand what should be quoted and what is an identifier.

> mikro-orm | 5.3.0

This is really dated.

**Reply by JasperVandenberghen** [reply]
Looking at the entity (this is an abandoned project I just took over so I am still getting familiar with the codebase), there is already a property.
  
  I'll make sure to bump up the deps.

**Reply by B4nan** [reply]
What query do you get if you fix the aliases as I suggested originally? As if the error is the same, it means the query is wrong.

**Reply by B4nan** [reply, marked as answer]
This is caused by the pagination mechanism, not sure why your initial query didn't have that (it should). Try to disable that:

```
qb.setFlag(QueryFlag.DISABLE_PAGINATE)
```

The results won't be correct tho, as without pagination subqueries in place, your limit is applied to the cartesian product - you join to many relations, which results in more rows being returned.

**Reply by B4nan** [reply]
Why are you joining those? They are not used in your query anyhow if I see it correctly. That's why the pagination is triggered.

```
        .leftJoin('o.users', 'ou')
        .leftJoin('ou.user', 'ouu')
```

**Reply by JasperVandenberghen** [reply]
```
 if (query.organizer) {
      void qb.andWhere({ [`ouu.id`]: query.organizer })
    }
```
This is a part below the code I posted. I could just create a seperate qb in the location and radius if statement that doesn't need those actually so I don't need the flag anymore. Thanks,!
